DAC CSV Export

elipe17 opened the pull request:
## Summary of Changes
- Added support for multi-selection list filters
- Added new filters for STT code, and fiscal year
- Added support for CSV export
- Made Records read only in DAC
- Updated `Newest` filter to display the newest records associated to each STT instead of to the latest datafile.
Pull request closes #2908 

## How to Test
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up -- build
```

1. Open http://localhost:3000/ and sign in as a user with admin privileges and permissions to search indexes.
2. Submit some files as different STTs
3. View the records and verify that records for the latest datafile for the respective STTs are shown
4. Export some or all of the records as CSV
5. Verify all records are in CSV, STT has its own column, and you can freely navigate DAC while the download is occurring

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] `DIGIT team/OFA system admin/OFA admin` can export parsed data via DAC to CSV
- [x] CSV export includes field names and STT code
- [x] CSV export only includes latest parsed data for any/all STTs 
- [x] anyone with access to search_indexes cannot edit/delete parsed data via DAC
- [x] filtering capability built to abovementioned specs
   - _in-place but needs refinement_ 
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Add mixin class/file to export admin data as csv

elipe17 committed:
- Add stt filter
- Rename file to filters.py

elipe17 committed:
- Add fiscal year filter

elipe17 committed:
- Add filters to models
- Update filter name

elipe17 committed:
- Convert to streaming interface to avoid timeouts

elipe17 committed:
- Move class to be internal

elipe17 committed:
- Enable export for all record types
- Add custom iterators to generate custom rows in the csv

elipe17 committed:
- fix liint

elipe17 committed:
- Add multiselector listfilter
- Update admin models

elipe17 committed:
- Rename Mixin
- Create base Mixin to appeal to MRO

elipe17 committed:
- Remove unused includes

elipe17 committed:
- lint

elipe17 committed:
- Order by stt_code

elipe17 committed:
- Make admin read only

elipe17 committed:
- Removing unnecessary method

elipe17 committed:
Merge branch 'develop' into dac-csv-export

elipe17 committed:
- fix lint

elipe17 committed:
Merge branch 'dac-csv-export' of https://github.com/raft-tech/TANF-app into dac-csv-export

andrew-jameson commented on tdrs-backend/tdpservice/search_indexes/admin/mixins.py:
```diff
@@ -0,0 +1,98 @@
+        return obj.datafile.stt.stt_code
+
+
+class AdminModelMixin(admin.ModelAdmin):
```
Is `ReadOnlyAdmin` the only child class? Why not just `ReadOnlyAdmin(admin.ModelAdmin)`?

elipe17 replied:
We need to have a common base class for all admin classes to subclass off of. If we dont, we break Python's MRO and crash. This is just a convenience way to help do that in the case that we need to add more mixins for our search index `ModelAdmin` classes. It doesnt work if you have multiple classes that inherit from `admin.ModelAdmin` and then have our search index `ModelAdmin` class inherit from them. Thats when you see the MRO error. We could do that if Python supported virtual inheritance like C++ does.

raftmsohani commented on tdrs-backend/tdpservice/search_indexes/admin/mixins.py:
```diff
@@ -0,0 +1,98 @@
+    pass
+
+
+class ReadOnlyAdmin(AdminModelMixin):
```
To limit repeating yourself, you should be able to add:
```
actions = ["export_as_csv"]
ordering = ['datafile__stt__stt_code']
```
here in this base class, however I think list_display and list_filter are different for each subclass. However, you might be able to add the more repeated ones, and then override the edge cases.

elipe17 replied:
Agreed, I should be able to call `super` to override these base lists if I need. I can make the change to avoid the duplication.

elipe17 replied:
I made some more class updates to aggregate similar mixins to a singular child class. Let me know your thoughts.

elipe17 committed:
- Aggregate Mixin classes
- Aggregate overrides to respective classes

elipe17 committed:
- lint

raftmsohani reviewed (approved):
LGTM!

jtimpe reviewed (approved):
CSV export and all the filters work great!

elipe17 committed:
Merge branch 'develop' into 2908-dac-csv-export

ADPennington committed:
Merge branch 'develop' into 2908-dac-csv-export

elipe17 committed:
Merge branch 'develop' into 2908-dac-csv-export

elipe17 committed:
Merge branch 'develop' into 2908-dac-csv-export

ADPennington committed:
Merge branch 'develop' into 2908-dac-csv-export

ADPennington committed:
Merge branch 'develop' into 2908-dac-csv-export

ADPennington commented on tdrs-backend/tdpservice/search_indexes/admin/tanf.py:
```diff
@@ -1,9 +1,9 @@
+class TANF_T1Admin(ReadOnlyAdminMixin, CsvExportAdminMixin):
     """ModelAdmin class for parsed T1 data files."""
 
     list_display = [
```
case_number, county fips, zip code, and stratum can be dropped from the list display in DAC

![2915p1](https://github.com/raft-tech/TANF-app/assets/63075587/882db073-6977-451b-bf62-643a949b1697)

ADPennington commented on tdrs-backend/tdpservice/search_indexes/admin/tanf.py:
```diff
@@ -14,98 +14,116 @@ class TANF_T1Admin(admin.ModelAdmin):
+class TANF_T2Admin(ReadOnlyAdminMixin, CsvExportAdminMixin):
     """ModelAdmin class for parsed T2 data files."""
 
     list_display = [
```
case number can be dropped from list display.

ADPennington commented on tdrs-backend/tdpservice/search_indexes/admin/tanf.py:
```diff
@@ -14,98 +14,116 @@ class TANF_T1Admin(admin.ModelAdmin):
     list_display = [
         'RecordType',
         'RPT_MONTH_YEAR',
         'CASE_NUMBER',
```
case number can be dropped from list display

ADPennington commented on tdrs-backend/tdpservice/search_indexes/admin/tanf.py:
```diff
@@ -14,98 +14,116 @@ class TANF_T1Admin(admin.ModelAdmin):
     list_display = [
         'RecordType',
         'RPT_MONTH_YEAR',
         'CASE_NUMBER',
```
case number can be dropped from list display

ADPennington commented on tdrs-backend/tdpservice/search_indexes/admin/tanf.py:
```diff
@@ -14,98 +14,116 @@ class TANF_T1Admin(admin.ModelAdmin):
     list_display = [
         'RecordType',
         'RPT_MONTH_YEAR',
         'CASE_NUMBER',
```
case number can be dropped from list display

elipe17 committed:
- remove unnecessary fields from list display

ADPennington commented on tdrs-backend/tdpservice/search_indexes/admin/filters.py:
```diff
@@ -0,0 +1,137 @@
+
+    def lookups(self, request, model_admin):
+        """Available options in dropdown."""
+        current_year = datetime.date.today().year
```
I would expect fiscal period on backend to be consistent with  what is available in the frontend dropdown. the minimum is FY2021, Q1.

elipe17 committed:
- Update to go from 2021 to current year

elipe17 committed:
Merge branch 'develop' into 2908-dac-csv-export

ADPennington reviewed (commented):
@elipe17 left some testing notes and feedback [here](https://hackmd.io/Ban2OmoQT_unHqLLLngUpg?view#DAC-CSV-Export).

elipe17 committed:
- Fix display bug with newest filter
- Remove erroneous filter options
- Update fiscal period query generation to match frontend generation exactly

elipe17 committed:
Merge branch '2908-dac-csv-export' of https://github.com/raft-tech/TANF-app into 2908-dac-csv-export

elipe17 committed:
- fix lint

elipe17 committed:
- logic fix

elipe17 commented:
@ADPennington answering your remaining questions and comments here:

`
Excel has limit of 1mil records. this will likely impact the export of Section 1, T3 records by fiscal period.  Does that limit impact how many records can be exported to csv? It's unlikely we would try to open the csv in excel; rather, the more common scenario is to import the csv into SAS, Python, etc. is it feasible to add a test about exporting >1mil records to csv?
`
- Creating a csv file with more than 1 million rows will not cause problems in terms of file size supported by csv. It can be as big as our memory budget will allow. Excel truncates the file and only show 1M rows to avoid memory issues and user experience issues. I tested this locally with ~400k records and the file size was ~50MB for reference. I don't think we have a feasible way to generate a test that has 1M records as of right now. I know @jtimpe has researched this idea a bit. Perhaps he can comment?

`
STT filter option is a bit strange. is it feasible to have an STT combobox like we have elsewhere (e.g. user profile) but with the added capability of multiple selections?
`
- It is feasible, but I think that will scope creep this ticket. Perhaps we can write an enhancement ticket to address the format?

`
Can we add the STT type filter to all search indexes? only tribes submit tribal TANF record types, for example.
`
- The STT filter is enabled on all search indexes.

`
The fiscal period filter seems to be arbitrarily set relative to today's date, but should be consistent with frontend fiscal period options. I dont think the fiscal filtering options should change relative to today's date.
`
- I updated the year generation algorithm to be a one to one match with the frontend's algorithm

`
If we have fiscal period filter, we do not also need the RPT_MONTH_YEAR field to filter by.
`
- I removed all the erroneous field filters

elipe17 committed:
Merge branch 'develop' into 2908-dac-csv-export

ADPennington reviewed (approved):
great work @elipe17 🚀 

**Summary**
* OFA permission groups can export latest data records by fiscal period and record type to csv from DAC :rocket:
* enhancements needed
    * filtering capability is in-place but needs refinement. each filter option needs to be applied in sequence (#2802) and the STT filter would be more useful as a combobox with multi-select options (#2950)
    * because the data submissions can happen at any time, having the datetime and fiscal period in the default naming convention for the export would be useful for version control purposes (#2949)
    * clean-up tasks/discussions
      * the prod db likely needs to be wiped before OFA can use it since parsing/validation logic has improved over time
      * digit team will need a short guide to navigate DAC for exporting data. we can work with UX on this. cc: @reitermb @robgendron 


test notes [here](https://hackmd.io/Ban2OmoQT_unHqLLLngUpg?view#DAC-CSV-Export)

elipe17 committed:
Merge branch 'develop' into 2908-dac-csv-export
